### DISCUSSION TITLE
Conventional Commit 도입 제안

### DISCUSSION THREAD

**Discussion by Kirade** [8percent/styleguide#74] [Ideas]
[Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) 에 따른 커밋 메시지 작성 방식의 도입을 제안합니다.

저희는 [기존 스타일 가이드](https://github.com/8percent/styleguide/tree/master/Git#commit-message) 에 따라서 아래와 같은 방식으로 작성하고 있었습니다.

- ADD: 입금 기능 추가
- CHG: 입금 기능 변경
- FIX: 입금 기능 수정
- REF: 입금 기능 리팩토링
- DEL: 입금 기능 제거

이 방식은 메시지의 접두어로 사용되는 단어들의 폭이 위의 5가지로 제한되는 상태입니다. 또한 사람이 읽기에는 문제가 없지만 오픈소스 등에서 널리 사용되는 방식과 다르게 사용되고 있어 커밋 메시지를 통한 자동화 기능을 쉽게 사용하기 어렵습니다.

따라서 앞서 언급한 [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) 을 도입을 제안합니다. 이 기능을 통해서 현재 방식의 아쉬운 부분을 해소할 수 있습니다. 접두어 선택이 조금 더 자유로운 편이고 메시지 작성 규칙이 기존에 비해 정교하게 정해져 있어 일관된 스타일의 커밋 메시지를 작성하기 쉬워집니다. 이런 점을 활용하여 여러 자동화 기능을 쉽게 붙일 수 있습니다.

**Comment by Kirade** [comment]
[참고 자료]
pre-commit 을 활용해 conventional commit 을 체크하는 패키지
- https://github.com/compilerla/conventional-pre-commit

**Comment by donghyun0218** [comment]
사용에 동의합니다. 
기존의 커밋 메세지들의 말머리는 기준이 모호할 때 사람마다 적용하는 말머리가 달라서 의미가 많이 상실되었다고 느껴지네요.

제안주신 Conventional Commit 을 읽어봤을 때 조직적으로 합의가 필요한 부분들이 조금씩 있을 것으로 보입니다.(Breaking Changes 사용 여부, scope 사용 여부 및 방법)
Conventional Commit 을 사용하는 것으로 정해진다면 위와 관련해서 함께 논의 해보면 좋을 것 같습니다!

**Reply by Kirade** [reply]
넵, 저희가 프로젝트 버저닝을 하지 않고 있기 때문에 이와 관련된 것들은 사용하지 않아도 괜찮을것 같네요.
말씀대로 이후에 논의해보시죠.

**Comment by chankoo** [comment]
PR 읽을때도 히스토리 파악이 쉬워질듯하고, 팀이 일관된 스타일로 커밋을 관리할 수 있어 좋아보입니다.

**Comment by ParkHyeonChae** [comment]
![image](https://github.com/user-attachments/assets/d1bdad50-2a99-4317-81e8-d64cb6fa58ea)


과제테스트시 컨벤셔널커밋으로 커밋을 관리했던 기억이 나네요. 저는 django프로젝트에서 변경의 대상이 되는 APP을 기준으로 scope에 포함했었습니다.

**Comment by jseop-lim** [comment]
이미 아실 수도 있지만, 참고차 공유드립니다.

커밋 규칙이 복잡해져 까먹을 수 있으므로 프로젝트 경로에 .gitmessage 파일을 추가하고 아래 명령어를 추가하여 커밋할 때마다 에디터에서 확인할 수 있습니다.

```shell
git config --global commit.template .gitmessage
```

더 복잡하게는 commit할 때 강제로 검사할 수도 있으나, 자세한 방법은 기억이 안나네요. 커맨드로 git을 관리하지 않는 경우, Pycharm에서는 commit message 형식을 등록 가능합니다.

제가 예시로 쓰는 .gitmessage 내용입니다.
```
# <type>: <subject>
##### Subject 50 characters ################# -> |

# Body Message
######## Body 72 characters ####################################### -> |

# Issue Tracker Number or URL

# --- COMMIT END ---
# Type can be
#   feat    : new feature
#   fix     : bug fix
#   build   : dependency
#   refactor: refactoring production code
#   style   : formatting, missing semi colons, etc; no code change
#   docs    : changes to documentation
#   test    : adding or refactoring tests
#             no productin code change
#   chore   : updating grunt tasks etc
#             no production code change
# ------------------
# Remember me ~
#   Capitalize the subject line
#   Use the imperative mood in the subject line
#   Do not end the subject line with a period
#   Separate subject from body with a blank line
#   Use the body to explain what and why vs. how
#   Can use multiple lines with "-" for bullet points in body
# ------------------
```

**Reply by Kirade** [reply]
커밋 메시지의 템플릿도 설정해둘수 있군요. 공유 감사합니다.
